src/v/config/throughput_control_group.cc

```diff
@@ -82,20 +108,36 @@ operator<<(std::ostream& os, const throughput_control_group& tcg) {
       "{{group_name: {}, client_id: {}, throughput_limit_node_in_bps: {}, "
       "throughput_limit_node_out_bps: {}}}",
       tcg.name,
-      tcg.client_id_re ? tcg.client_id_re->pattern() : ""s,
+      tcg.client_id_matcher,
       tcg.throughput_limit_node_in_bps,
       tcg.throughput_limit_node_out_bps);
     return os;
 }
 
 bool throughput_control_group::match_client_id(
-  const ss::sstring& client_id) const {
-    if (!client_id_re) {
+  const std::optional<std::string_view> client_id) const {
+    if (!client_id_matcher) {
         // omitted match criterion means "always match"
         return true;
     }
-    return re2::RE2::FullMatch(
-      re2::StringPiece(client_id.c_str(), client_id.length()), *client_id_re);
+    return std::visit(
+      [client_id](auto&& v) -> bool {
+          using T = std::decay_t<decltype(v)>;
+          if constexpr (std::is_same_v<
+                          T,
+                          client_id_matcher_type::unspecified_tag>) {
+              // only missing client_id matches the unspecified tag
+              return !client_id;
+          } else if constexpr (std::is_same_v<T, copyable_RE2>) {
+              // missing client_id never matches a re
+              return client_id
+                     && re2::RE2::FullMatch(re2::StringPiece(*client_id), v);
+          } else {
+              static_assert(
+                always_false_v<T>, "non-exhaustive client_id_re_type visitor");
+          }
+      },
+      client_id_matcher->v);
```

Comment:
since this PR introduces throughput_control_group.cc why have this other commit that rewrites a bunch of it? seems like this should be squashed with the second commit to avoid extra reviewing?

Reply:
I was trying to split the large file into smaller parts to make reviewing easier :) If that's the other way around actually, I will squash

Reply:
yeh i mean i think it's fine the way it is. i only mentioned it because it seemed like enough code was changing that it wasn't necessarily making review easier. but it isn't a problem either way.